src/ESPEasy-Globals.h

```diff
@@ -306,6 +306,13 @@ void check_size() {
 #define CPLUGIN_INIT                       50
 #define CPLUGIN_UDP_IN                     51
 
+// new messages for autodiscover controller plugins (experimental) i.e. C014
+#define CPLUGIN_CONNECTED                  52 // call after connected to mqtt server to publich device autodicover features
+#define CPLUGIN_DICONNECT                  53 // should be called before major changes i.e. changing the device name to clean up data on the controller. !ToDo
+#define CPLUGIN_FALLING_ASLEEP             54 // call before falling asleep !ToDo
```

Comment:
I have recently added a similar function to perform a flush.
Is that the same behavior?

Reply:
Hmm, it has not yet been merged, but I meant this PR: https://github.com/letscontrolit/ESPEasy/pull/2426/files

Reply:
Let`s start with an easy one (hope I do this right, that's all new for me, please be patient)

I think CPLUGIN_FALLING_ASLEEP will do the same as to your CPLUGIN_FLUSH  they could be equal or I use CPLUGIN_FLUSH and delete CPLUGIN_FALLING_ASLEEP.

```suggestion
#define CPLUGIN_CONNECTED                  54 // call after connected to mqtt server to publich device autodicover features
#define CPLUGIN_FALLING_ASLEEP             52 // call before falling asleep
```

Reply:
I think CPLUGIN_FLUSH is more descriptive of what it does.
CPLUGIN_FALLING_ASLEEP is more about when it should be executed.

CPLUGIN_CONNECTED  could maybe also be renamed to `CPLUGIN_ON_CONNECT` or `CPLUGIN_GOT_CONNECTED`, which is more in line with the events used for WiFi connection states.

Reply:
OK, fine will use `CPLUGIN_FLUSH` and `CPLUGIN_GOT_CONNECTED` (or is there already a event after MQTT connected I can use?)

perhaps like this?
```
#define CPLUGIN_FLUSH                      52
// new messages for autodiscover controller plugins (experimental) i.e. C014
#define CPLUGIN_GOT_CONNECTED              53 // call after connected to mqtt server to publich device autodicover features
#define CPLUGIN_GOT_INVALID                54 // should be called before major changes i.e. changing the device name to clean up data on the controller. !ToDo
#define CPLUGIN_INTERVAL                   55 // call every interval loop
#define CPLUGIN_FEEDBACK                   56 // call for sending feedback !ToDo done by direct function call in PluginCall() for now.
```

Reply:
Not sure yet.
The other controllers do not have a lot of bi-directional communication.
So I don't think there is any other function that can be re-used for this.

Reply:
OK!?

so how to do the following tasks better?
**CONNECTED:** Homie (and perhaps other conventions) need to send information about the properties of the unit after reboot.
**INVALID:** to inform that the controller can't expect further data under the this topic homie/%unitName%/#. i.e. when the unit name changes.
**INTERVAL:** to send a heartbeat. Currently ESPEasy works by a delay sending data every X seconds. That is normaly heartbeat enough. Homie 4 (development) do not need interval Information anymore. Homie 3 wants signal strength and uptime.

**FEEDBACK:** It is essential (for Homie) that after message to .../%valueName%/set a feedback to .../%valueName% is sent. A switch in openHAB stays in an intermediate state until it receives the feedback. Additionally i.e. GPIO changes (by a rule or other command source) should be sent to the controller to keep the values up to date). Currently a json formatted feedback is sent after a every mqtt comand setting a GPIO value which could/should be integrated into the corresponding controller.

Initially I thought that all functions which are specific to a controller should be handled by the controller plugin via these function IDs. I thought they do not harm any others. In the case of the json message implemented outside the plugin I can imagine controllers triggering errors (and openHAB is quite fast with "communication errors") when receiving incorrect messages.

Reply:
Sounds good
Only the "feedback" I'm not sure about.
What about "acknowledge" ?

Reply:
changed to CPLUGIN_ACKNOWLEDGE.